File: engine/packages/guard/src/routing/actor_path.rs
use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use gas::prelude::*;
use rivet_types::actors::CrashPolicy;
use serde::Deserialize;

use crate::errors;

/// The `rvt-` query parameter prefix is reserved for Rivet gateway routing.
/// All query parameters with this prefix are stripped before forwarding
/// requests to the actor, so actors will never see them.
const RVT_PREFIX: &str = "rvt-";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ActorPathInfo {
	pub actor_id: String,
	pub token: Option<String>,
	pub stripped_path: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct QueryActorPathInfo {
	pub query: QueryActorQuery,
	pub token: Option<String>,
	pub stripped_path: String,
}

#[derive(Debug, Clone, PartialEq)]
pub enum QueryActorQuery {
	Get {
		namespace: String,
		name: String,
		key: Vec<String>,
	},
	GetOrCreate {
		namespace: String,
		name: String,
		pool_name: String,
		key: Vec<String>,
		input: Option<Vec<u8>>,
		region: Option<String>,
		crash_policy: Option<CrashPolicy>,
	},
}

#[derive(Debug, Clone)]
pub enum ParsedActorPath {
	Direct(ActorPathInfo),
	Query(QueryActorPathInfo),
}

/// Parsed rvt-* query parameters.
#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
struct RvtParams {
	namespace: String,
	method: String,
	#[serde(default)]
	runner: Option<String>,
	#[serde(default)]
	pool: Option<String>,
	#[serde(default)]
	key: Option<String>,
	#[serde(default)]
	input: Option<String>,
	#[serde(default)]
	region: Option<String>,
	#[serde(default, rename = "crash-policy")]
	crash_policy: Option<String>,
	#[serde(default)]
	token: Option<String>,
}

/// Parse actor routing information from path.
/// Matches patterns:
/// - /gateway/{actor_id}/{...path}
/// - /gateway/{actor_id}@{token}/{...path}
/// - /gateway/{name}/{...path}?rvt-namespace=...&rvt-method=...
///
/// Returns `Ok(None)` for paths that are not gateway paths or for malformed
/// direct paths (backwards-compatible silent fallthrough). Returns `Err` for
/// malformed query paths so clients receive actionable error messages.
pub fn parse_actor_path(path: &str) -> Result<Option<ParsedActorPath>> {
	// Extract base path and raw query from the original path string directly,
	// without running through a URL parser, to preserve actor query params
	// byte-for-byte (no re-encoding of %20, +, etc.).
	let (base_path, raw_query) = split_path_and_query(path);

	if base_path.contains("//") {
		return Ok(None);
	}

	let segments: Vec<&str> = base_path
		.split('/')
		.filter(|segment| !segment.is_empty())
		.collect();
	if segments.first().copied() != Some("gateway") {
		return Ok(None);
	}

	let raw_query_str = raw_query.unwrap_or("");

	// Check if any raw query param key starts with rvt-.
	let has_rvt = !raw_query_str.is_empty()
		&& raw_query_str
			.split('&')
			.any(|part| part.split('=').next().unwrap_or("").starts_with(RVT_PREFIX));

	if has_rvt {
		let rvt_params = extract_rvt_params_from_raw_query(raw_query_str)?;
		let actor_query_string = strip_rvt_query_params(raw_query_str);
		return parse_query_actor_path(base_path, &segments, &rvt_params, &actor_query_string)
			.map(|path| Some(ParsedActorPath::Query(path)));
	}

	// Direct path: pass the raw query string through unchanged.
	let raw_query_string = match raw_query {
		Some(q) => format!("?{q}"),
		None => String::new(),
	};
	Ok(parse_direct_actor_path(
		base_path,
		&segments,
		&raw_query_string,
	))
}

fn parse_direct_actor_path(
	base_path: &str,
	segments: &[&str],
	raw_query_string: &str,
) -> Option<ParsedActorPath> {
	if segments.len() < 2 || segments[0] != "gateway" {
		return None;
	}

	let actor_segment = segments[1];
	if actor_segment.is_empty() {
		return None;
	}

	let (actor_id, token) = if let Some(at_pos) = actor_segment.find('@') {
		let raw_actor_id = &actor_segment[..at_pos];
		let raw_token = &actor_segment[at_pos + 1..];
		if raw_actor_id.is_empty() || raw_token.is_empty() {
			return None;
		}

		let actor_id = urlencoding::decode(raw_actor_id).ok()?.into_owned();
		let token = urlencoding::decode(raw_token).ok()?.into_owned();
		(actor_id, Some(token))
	} else {
		(urlencoding::decode(actor_segment).ok()?.into_owned(), None)
	};

	let remaining_path = build_remaining_path(base_path, raw_query_string, 2);

	Some(ParsedActorPath::Direct(ActorPathInfo {
		actor_id,
		token,
		stripped_path: remaining_path,
	}))
}

fn parse_query_actor_path(
	base_path: &str,
	segments: &[&str],
	rvt_params: &[(String, String)],
	actor_query_string: &str,
) -> Result<QueryActorPathInfo> {
	let name_segment = segments
		.get(1)
		.copied()
		.ok_or_else(|| errors::QueryEmptyActorName.build())?;

	if name_segment.contains('@') {
		return Err(errors::QueryPathTokenSyntax.build());
	}

	let name = urlencoding::decode(name_segment).map_err(|_| {
		errors::QueryInvalidPercentEncoding {
			name: "name".to_string(),
		}
		.build()
	})?;

	if name.is_empty() {
		return Err(errors::QueryEmptyActorName.build());
	}

	let rvt = extract_rvt_params(rvt_params)?;
	let stripped_path = build_remaining_path(base_path, actor_query_string, 2);

	Ok(QueryActorPathInfo {
		token: rvt.token.clone(),
		query: build_actor_query(&name, rvt)?,
		stripped_path,
	})
}

/// Extract and validate rvt-* params from pre-parsed query pairs.
fn extract_rvt_params(rvt_params: &[(String, String)]) -> Result<RvtParams> {
	let mut map = serde_json::Map::new();

	for (raw_key, value) in rvt_params {
		let stripped = raw_key
			.strip_prefix(RVT_PREFIX)
			.expect("rvt_params should only contain rvt- prefixed keys");

		if map.contains_key(stripped) {
			return Err(errors::QueryDuplicateParam {
				name: raw_key.clone(),
			}
			.build());
		}
		map.insert(
			stripped.to_string(),
			serde_json::Value::String(value.clone()),
		);
	}

	serde_json::from_value(serde_json::Value::Object(map)).map_err(|e| {
		errors::QueryInvalidParams {
			detail: e.to_string(),
		}
		.build()
	})
}

/// Split a comma-separated key string into components.
/// Missing or empty key yields an empty vec.
fn split_key(raw: Option<&str>) -> Vec<String> {
	match raw {
		None | Some("") => Vec::new(),
		Some(s) => s.split(',').map(String::from).collect(),
	}
}

fn build_actor_query(name: &str, rvt: RvtParams) -> Result<QueryActorQuery> {
	let key = split_key(rvt.key.as_deref());

	match rvt.method.as_str() {
		"get" => {
			if rvt.input.is_some()
				|| rvt.region.is_some()
				|| rvt.crash_policy.is_some()
				|| rvt.runner.is_some()
			{
				return Err(errors::QueryGetDisallowedParams.build());
			}

			Ok(QueryActorQuery::Get {
				namespace: rvt.namespace,
				name: name.to_string(),
				key,
			})
		}
		"getOrCreate" => {
			let pool_name = rvt
				.pool
				.or(rvt.runner)
				.ok_or_else(|| errors::QueryMissingPool.build())?;

			let input = rvt.input.as_deref().map(decode_query_input).transpose()?;

			let crash_policy = rvt
				.crash_policy
				.as_deref()
				.map(parse_crash_policy)
				.transpose()?;

			Ok(QueryActorQuery::GetOrCreate {
				namespace: rvt.namespace,
				name: name.to_string(),
				pool_name,
				key,
				input,
				region: rvt.region,
				crash_policy,
			})
		}
		other => Err(errors::QueryInvalidParams {
			detail: format!("unknown method: {other}, expected get or getOrCreate"),
		}
		.build()),
	}
}

fn parse_crash_policy(value: &str) -> Result<CrashPolicy> {
	match value {
		"restart" => Ok(CrashPolicy::Restart),
		"sleep" => Ok(CrashPolicy::Sleep),
		"destroy" => Ok(CrashPolicy::Destroy),
		other => Err(errors::QueryInvalidParams {
			detail: format!("unknown crash policy: {other}, expected restart, sleep, or destroy"),
		}
		.build()),
	}
}

fn decode_query_input(value: &str) -> Result<Vec<u8>> {
	if !value
		.bytes()
		.all(|byte| byte.is_ascii_alphanumeric() || byte == b'-' || byte == b'_')
		|| value.len() % 4 == 1
	{
		return Err(errors::QueryInvalidBase64Input.build());
	}

	let bytes = URL_SAFE_NO_PAD
		.decode(value.as_bytes())
		.map_err(|_| errors::QueryInvalidBase64Input.build())?;

	validate_cbor(&bytes).map_err(|err| {
		errors::QueryInvalidCborInput {
			detail: err.to_string(),
		}
		.build()
	})?;

	Ok(bytes)
}

/// Split a path into the base path and the raw query string (without `?`).
/// Fragments are stripped.
fn split_path_and_query(path: &str) -> (&str, Option<&str>) {
	let path = match path.find('#') {
		Some(pos) => &path[..pos],
		None => path,
	};
	match path.find('?') {
		Some(pos) => (&path[..pos], Some(&path[pos + 1..])),
		None => (path, None),
	}
}

/// Extract rvt-* params from a raw query string, decoding their values
/// using form-urlencoded rules (`+` as space, then percent-decode).
fn extract_rvt_params_from_raw_query(raw_query: &str) -> Result<Vec<(String, String)>> {
	let mut params = Vec::new();
	for part in raw_query.split('&') {
		let (raw_key, raw_value) = match part.find('=') {
			Some(pos) => (&part[..pos], &part[pos + 1..]),
			None => (part, ""),
		};
		if raw_key.starts_with(RVT_PREFIX) {
			let decoded_value = decode_form_value(raw_value).map_err(|_| {
				errors::QueryInvalidPercentEncoding {
					name: raw_key.to_string(),
				}
				.build()
			})?;
			params.push((raw_key.to_string(), decoded_value));
		}
	}
	Ok(params)
}

/// Decode a form-urlencoded value: treat `+` as space, then percent-decode.
fn decode_form_value(raw: &str) -> std::result::Result<String, std::string::FromUtf8Error> {
	let with_spaces = raw.replace('+', " ");
	urlencoding::decode(&with_spaces).map(|s| s.into_owned())
}

/// Strip rvt-* params from a raw query string, preserving actor params
/// byte-for-byte without re-encoding.
fn strip_rvt_query_params(raw_query: &str) -> String {
	let actor_parts: Vec<&str> = raw_query
		.split('&')
		.filter(|part| {
			if part.is_empty() {
				return false;
			}
			let key = part.split('=').next().unwrap_or("");
			!key.starts_with(RVT_PREFIX)
		})
		.collect();

	if actor_parts.is_empty() {
		String::new()
	} else {
		format!("?{}", actor_parts.join("&"))
	}
}

fn build_remaining_path(base_path: &str, query_string: &str, consumed_segments: usize) -> String {
	let segments: Vec<&str> = base_path
		.split('/')
		.filter(|segment| !segment.is_empty())
		.collect();

	let mut prefix_len = 0;
	for segment in segments.iter().take(consumed_segments) {
		prefix_len += 1 + segment.len();
	}

	let remaining_base = if prefix_len < base_path.len() {
		&base_path[prefix_len..]
	} else {
		"/"
	};

	if remaining_base.is_empty() || !remaining_base.starts_with('/') {
		format!("/{remaining_base}{query_string}")
	} else {
		format!("{remaining_base}{query_string}")
	}
}

fn validate_cbor(bytes: &[u8]) -> Result<()> {
	ciborium::from_reader::<ciborium::Value, _>(bytes).context("invalid cbor")?;
	Ok(())
}

File: rivetkit-typescript/packages/rivetkit/src/actor-gateway/actor-path.ts
/**
 * Actor gateway path parsing.
 *
 * Parses `/gateway/{...}` paths into either a direct actor ID path or a query
 * path with rvt-* query parameters. This is the TypeScript equivalent of the engine
 * parser at `engine/packages/guard/src/routing/actor_path.rs`.
 */
import * as cbor from "cbor-x";
import * as errors from "@/actor/errors";
import {
	type ActorGatewayQuery,
	type CrashPolicy,
	GetForKeyRequestSchema,
	GetOrCreateRequestSchema,
} from "@/client/query";

/**
 * The `rvt-` query parameter prefix is reserved for Rivet gateway routing.
 * All query parameters with this prefix are stripped before forwarding
 * requests to the actor, so actors will never see them.
 */
const RVT_PREFIX = "rvt-";

/**
 * A direct actor path targets a specific actor by its ID.
 * Format: `/gateway/{actorId}[@{token}]/{...path}`
 *
 * The actor ID is extracted directly from the URL and no query resolution is
 * needed. This is the path format used by resolved handles and connections
 * that already know their target actor.
 */
export interface ParsedDirectActorPath {
	type: "direct";
	actorId: string;
	token?: string;
	remainingPath: string;
}

/**
 * A query actor path resolves to an actor via a key-based lookup.
 * Format: `/gateway/{name}/{...path}?rvt-namespace=...&rvt-method=get|getOrCreate&rvt-key=...`
 *
 * The actor name is a clean path segment, and all routing params are rvt-*
 * query parameters that get stripped before forwarding to the actor. This path
 * must be resolved to a concrete actor ID before proxying, using the engine
 * control client's getWithKey or getOrCreateWithKey methods.
 *
 * This is the engine-side reference implementation's TypeScript equivalent.
 * See `engine/packages/guard/src/routing/actor_path.rs` for the Rust counterpart.
 */
export interface ParsedQueryActorPath {
	type: "query";
	query: ActorGatewayQuery;
	namespace: string;
	runnerName?: string;
	crashPolicy?: CrashPolicy;
	token?: string;
	remainingPath: string;
}

export type ParsedActorPath = ParsedDirectActorPath | ParsedQueryActorPath;

/**
 * Parse actor routing information from a gateway path.
 *
 * Returns a `ParsedDirectActorPath` or `ParsedQueryActorPath` depending on the
 * URL structure, or `null` if the path does not start with `/gateway/`.
 *
 * Detection heuristic: if any query parameter starts with `rvt-`, it is a query
 * path. Otherwise it is a direct actor ID path. The two cases are handled by
 * `parseQueryActorPath` and `parseDirectActorPath` respectively.
 *
 * This must stay in sync with the engine parser at
 * `engine/packages/guard/src/routing/actor_path.rs`.
 */
export function parseActorPath(path: string): ParsedActorPath | null {
	// Extract base path and raw query from the original path string directly,
	// without running through a URL parser, to preserve actor query params
	// byte-for-byte (no re-encoding of %20, +, etc.).
	const [basePath, rawQuery] = splitPathAndQuery(path);

	if (basePath.includes("//")) {
		return null;
	}

	const segments = basePath.split("/").filter((s) => s.length > 0);
	if (segments[0] !== "gateway") {
		return null;
	}

	const rawQueryStr = rawQuery ?? "";

	// Check if any raw query param key starts with rvt-.
	const hasRvt =
		rawQueryStr.length > 0 &&
		rawQueryStr.split("&").some((part) => {
			const key = part.split("=")[0];
			return key.startsWith(RVT_PREFIX);
		});

	if (hasRvt) {
		const rvtParams = extractRvtParamsFromRaw(rawQueryStr);
		const actorQueryString = stripRvtQueryParams(rawQueryStr);
		return parseQueryActorPath(
			basePath,
			segments,
			rvtParams,
			actorQueryString,
		);
	}

	// Direct path: pass the raw query string through unchanged.
	const rawQueryString = rawQuery !== null ? `?${rawQuery}` : "";
	return parseDirectActorPath(basePath, rawQueryString);
}

function parseDirectActorPath(
	basePath: string,
	rawQueryString: string,
): ParsedDirectActorPath | null {
	const segments = basePath.split("/").filter((s) => s.length > 0);

	if (segments.length < 2 || segments[0] !== "gateway") {
		return null;
	}

	const actorSegment = segments[1];
	if (actorSegment.length === 0) {
		return null;
	}

	let actorId: string;
	let token: string | undefined;

	const atPos = actorSegment.indexOf("@");
	if (atPos !== -1) {
		const rawActorId = actorSegment.slice(0, atPos);
		const rawToken = actorSegment.slice(atPos + 1);

		if (rawActorId.length === 0 || rawToken.length === 0) {
			return null;
		}

		try {
			actorId = decodeURIComponent(rawActorId);
			token = decodeURIComponent(rawToken);
		} catch {
			return null;
		}
	} else {
		try {
			actorId = decodeURIComponent(actorSegment);
		} catch {
			return null;
		}
		token = undefined;
	}

	const remainingPath = buildRemainingPath(basePath, rawQueryString, 2);

	return {
		type: "direct",
		actorId,
		token,
		remainingPath,
	};
}

function parseQueryActorPath(
	basePath: string,
	segments: string[],
	rvtParams: Array<[string, string]>,
	actorQueryString: string,
): ParsedQueryActorPath {
	const nameSegment = segments[1];
	if (!nameSegment || nameSegment.length === 0) {
		throw new errors.InvalidRequest(
			"query gateway actor name must not be empty",
		);
	}

	if (nameSegment.includes("@")) {
		throw new errors.InvalidRequest(
			"query gateway paths must not use @token syntax",
		);
	}

	let name: string;
	try {
		name = decodeURIComponent(nameSegment);
	} catch {
		throw new errors.InvalidRequest(
			"invalid percent-encoding for query gateway param 'name'",
		);
	}

	if (name.length === 0) {
		throw new errors.InvalidRequest(
			"query gateway actor name must not be empty",
		);
	}

	const rvt = extractRvtParams(rvtParams);
	const remainingPath = buildRemainingPath(basePath, actorQueryString, 2);

	return {
		type: "query",
		query: buildActorQuery(name, rvt),
		namespace: rvt.namespace,
		runnerName: rvt.runner,
		crashPolicy: rvt.crashPolicy,
		token: rvt.token,
		remainingPath,
	};
}

interface RvtParams {
	namespace: string;
	method: string;
	runner?: string;
	key: string[];
	input?: unknown;
	region?: string;
	crashPolicy?: CrashPolicy;
	token?: string;
}

function splitKey(raw: string | undefined): string[] {
	if (raw === undefined || raw === "") return [];
	return raw.split(",");
}

function extractRvtParams(rvtRaw: Array<[string, string]>): RvtParams {
	const params = new Map<string, string>();

	for (const [rawKey, value] of rvtRaw) {
		const stripped = rawKey.slice(RVT_PREFIX.length);

		if (
			stripped === "namespace" ||
			stripped === "method" ||
			stripped === "runner" ||
			stripped === "key" ||
			stripped === "input" ||
			stripped === "region" ||
			stripped === "crash-policy" ||
			stripped === "token"
		) {
			if (params.has(stripped)) {
				throw new errors.InvalidRequest(
					`duplicate query gateway param: ${rawKey}`,
				);
			}
			params.set(stripped, value);
		} else {
			throw new errors.InvalidRequest(
				`unknown query gateway param: ${rawKey}`,
			);
		}
	}

	const requireParam = (name: string): string => {
		const value = params.get(name);
		if (value === undefined) {
			throw new errors.InvalidRequest(
				`missing required param: rvt-${name}`,
			);
		}
		return value;
	};

	const namespace = requireParam("namespace");
	const method = requireParam("method");
	const runner = params.get("runner");
	const key = splitKey(params.get("key"));
	const region = params.get("region");
	const token = params.get("token");

	// Decode input CBOR if present.
	const inputRaw = params.get("input");
	let input: unknown;
	if (inputRaw !== undefined) {
		const inputBuffer = decodeBase64Url(inputRaw);
		try {
			input = cbor.decode(inputBuffer);
		} catch (cause) {
			throw new errors.InvalidRequest(
				`invalid query gateway input cbor: ${cause}`,
			);
		}
	}

	// Parse crash policy.
	const crashPolicyRaw = params.get("crash-policy");
	let crashPolicy: CrashPolicy | undefined;
	if (crashPolicyRaw !== undefined) {
		if (
			crashPolicyRaw !== "restart" &&
			crashPolicyRaw !== "sleep" &&
			crashPolicyRaw !== "destroy"
		) {
			throw new errors.InvalidRequest(
				`unknown crash policy: ${crashPolicyRaw}, expected restart, sleep, or destroy`,
			);
		}
		crashPolicy = crashPolicyRaw;
	}

	return {
		namespace,
		method,
		runner,
		key,
		input,
		region,
		crashPolicy,
		token,
	};
}

function buildActorQuery(name: string, rvt: RvtParams): ActorGatewayQuery {
	switch (rvt.method) {
		case "get": {
			if (
				rvt.input !== undefined ||
				rvt.region !== undefined ||
				rvt.crashPolicy !== undefined ||
				rvt.runner !== undefined
			) {
				throw new errors.InvalidRequest(
					"query gateway method=get does not allow rvt-input, rvt-region, rvt-crash-policy, or rvt-runner params",
				);
			}

			return {
				getForKey: GetForKeyRequestSchema.parse({
					name,
					key: rvt.key,
				}),
			};
		}
		case "getOrCreate": {
			if (rvt.runner === undefined) {
				throw new errors.InvalidRequest(
					"query gateway method=getOrCreate requires rvt-runner param",
				);
			}

			return {
				getOrCreateForKey: GetOrCreateRequestSchema.parse({
					name,
					key: rvt.key,
					input: rvt.input,
					region: rvt.region,
				}),
			};
		}
		default:
			throw new errors.InvalidRequest(
				`unknown method: ${rvt.method}, expected get or getOrCreate`,
			);
	}
}

function decodeBase64Url(value: string): Uint8Array {
	if (!/^[A-Za-z0-9_-]*$/.test(value) || value.length % 4 === 1) {
		throw new errors.InvalidRequest(
			"invalid base64url in query gateway input",
		);
	}

	const paddingLength = (4 - (value.length % 4 || 4)) % 4;
	const base64 =
		value.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat(paddingLength);

	if (typeof Buffer !== "undefined") {
		return new Uint8Array(Buffer.from(base64, "base64"));
	}

	const binary = atob(base64);
	const buffer = new Uint8Array(binary.length);
	for (let i = 0; i < binary.length; i++) {
		buffer[i] = binary.charCodeAt(i);
	}

	return buffer;
}

/** Split a path into the base path and the raw query string (without `?`). Fragments are stripped. */
function splitPathAndQuery(path: string): [string, string | null] {
	const fragmentPos = path.indexOf("#");
	const pathNoFragment =
		fragmentPos !== -1 ? path.slice(0, fragmentPos) : path;
	const queryPos = pathNoFragment.indexOf("?");
	if (queryPos !== -1) {
		return [
			pathNoFragment.slice(0, queryPos),
			pathNoFragment.slice(queryPos + 1),
		];
	}
	return [pathNoFragment, null];
}

/**
 * Extract rvt-* params from a raw query string, decoding their values
 * using form-urlencoded rules (`+` as space, then percent-decode).
 */
function extractRvtParamsFromRaw(rawQuery: string): Array<[string, string]> {
	const params: Array<[string, string]> = [];
	for (const part of rawQuery.split("&")) {
		const eqPos = part.indexOf("=");
		const rawKey = eqPos !== -1 ? part.slice(0, eqPos) : part;
		const rawValue = eqPos !== -1 ? part.slice(eqPos + 1) : "";

		if (rawKey.startsWith(RVT_PREFIX)) {
			let decodedValue: string;
			try {
				decodedValue = decodeFormValue(rawValue);
			} catch {
				throw new errors.InvalidRequest(
					`invalid percent-encoding for query gateway param '${rawKey}'`,
				);
			}
			params.push([rawKey, decodedValue]);
		}
	}
	return params;
}

/** Decode a form-urlencoded value: treat `+` as space, then percent-decode. */
function decodeFormValue(raw: string): string {
	return decodeURIComponent(raw.replace(/\+/g, " "));
}

/**
 * Strip rvt-* params from a raw query string, preserving actor params
 * byte-for-byte without re-encoding.
 */
function stripRvtQueryParams(rawQuery: string): string {
	const actorParts = rawQuery.split("&").filter((part) => {
		if (part.length === 0) return false;
		const key = part.split("=")[0];
		return !key.startsWith(RVT_PREFIX);
	});
	return actorParts.length === 0 ? "" : `?${actorParts.join("&")}`;
}

function buildRemainingPath(
	basePath: string,
	queryString: string,
	consumedSegments: number,
): string {
	const segments = basePath
		.split("/")
		.filter((segment) => segment.length > 0);

	let prefixLen = 0;
	for (let i = 0; i < consumedSegments; i++) {
		prefixLen += 1 + segments[i].length;
	}

	let remainingBase: string;
	if (prefixLen < basePath.length) {
		remainingBase = basePath.slice(prefixLen);
	} else {
		remainingBase = "/";
	}

	if (remainingBase.length === 0 || !remainingBase.startsWith("/")) {
		return `/${remainingBase}${queryString}`;
	}

	return `${remainingBase}${queryString}`;
}
